### ISSUE TITLE
Mypy does not correctly narrow indexing operations when using Literal or Final keys

### ISSUE THREAD

**Issue by Michael0x2a** [python/mypy#7905]
Consider the following program:

```
from typing_extensions import Literal, TypedDict
from enum import Enum

class Key(Enum):
    X = 1
    Y = 2
    Z = 3

class MyDict(TypedDict):
    key: Literal[Key.X, Key.Y]
    blah: int

KEY: Literal["key"] = "key"

d: MyDict

if d["key"] is Key.X:
    reveal_type(d["key"])  # note: Revealed type is 'Literal[Key.X]'

if d[KEY] is Key.X:
    reveal_type(d[KEY])  # note: Revealed type is 'Literal[Key.X, Key.Y]'

```

Mypy is currently capable of narrowing expressions like `d["key"]`, which we can see in the first expression.

So, it's natural to assume that mypy would be able to do the same for the second since the two programs are theoretically identical -- but we can't.

The root cause has to do with the "literal" subsystem (which is not to be confused with the Literal types subsystem) here: https://github.com/python/mypy/blob/master/mypy/literals.py#L65

The index in the second example is a NameExpr, which causes the if statement to evaluate to false and return a `LITERAL_NO`. This then makes the narrowing logic rule out `d[KEY]` as a candidate for narrowing in https://github.com/python/mypy/blob/master/mypy/checker.py#L3775.

I'm not really sure what the best way of fixing this would be. The natural solution would be to also pass along the expression type into the `literal(...)` function, but that seems very annoying to do. I'm also not entirely sure whether this is even a sound narrowing: I'm not very familiar with the "literals" subsystem, or how it's meant to interact with Literal types.

**Comment by ilevkivskyi** [comment]
Does this happen often? If not, I would propose to not worry about this for now (mostly because I am not sure what to do either).

**Comment by ilevkivskyi** [comment]
Related https://github.com/python/mypy/issues/7339 (even stronger version of this).

**Comment by Michael0x2a** [comment]
Yeah, it's somewhat low priority. I mostly added this issue because it came up while I was working on another PR and wanted something to link to.

**Comment by alicederyn** [comment]
I believe this has been resolved?

https://mypy-play.net/?gist=8745b8ac7eb7280426ab2e5ed7451c91

Now shows `Revealed type is "Literal[__main__.Key.X] | Literal[__main__.Key.Y]"`, which could be simplified but is technically correct.

**Comment by ilevkivskyi** [comment]
IIUC @Michael0x2a wants both revealed typed to be `Literal[__main__.Key.X]`. I am not sure I like the idea, but I am also not against it either.
